### ISSUE TITLE
CMake build infrastructure

### ISSUE THREAD

**Issue by TheAssassin** [AppImage/AppImageKit#408]
It would be nice to have a modern and reliable build infrastructure. Thus, I'd like to see a CMake build infrastructure.

**Comment by probonopd** [comment]
I won't do it (I was happy enough to get it to build exactly in the way I need using bash) but if someone would like to try, please go ahead. Only request: Work in a private branch please and only send a PR once everything is working as it is now, including signatures, updateinformation, etc.

**Comment by TheAssassin** [comment]
You're not the "I have a workplace and build and test my code locally" type of guy, so I understand that you are happy with that bash infrastructure, as you just have to let Travis build it for you. Though, there's several points that have to be improved to support e.g. incremental builds, including a better handling of external dependencies.

I opened this issue primarily to remind me of doing that one day.

**Comment by probonopd** [comment]
Occasionally I do build locally, then I copy snippets from the bash script and run them on the shell. So I can definitely see the room for improvement there ;-)

**Comment by tresf** [comment]
Hope these are relevant, but there are some tools to help with dependency management.  A colleague provided these to me researching dependency management for (currently private) project... 

Hunter https://github.com/ruslo/hunter
> "I find this one to be very interesting as it's written entirely in CMake scripts, which means there's no other dependencies (it could even download the compiler for you, although there's only Clang atm). It's used by including a single file HunterGate.cmake, which automatically downloads Hunter itself. Because it's so tightly integrated with the CMake build process, it will automatically detect your current compiler and download/build a version that's ABI compatible."


https://github.com/conan-io/conan
Conan C/C++ Package Manager
> "Written in Python. Seems to be better documented than Hunter. Provides a "cmake generator" which outputs a file that you can include in CMake in order to easily use the installed packages."


CPM https://github.com/iauns/cpm
> "Also written in CMake. It claims to allow "you to link against multiple different versions of the same static library without symbol conflicts". I've yet to take a closer look at it."

**Comment by TheAssassin** [comment]
Actually, dependency management is not really an issue I guess. The current build script is capable of fetching and building dependencies, too. It's not very clean, and rather complicated to port, but it's doable. Please see https://github.com/TheAssassin/zsync-curl/blob/master/CMakeLists.txt for instance. The CMake configuration is a port of the legacy build script.

**Comment by tresf** [comment]
> Actually, dependency management is not really an issue I guess. The current build script is capable of fetching and building dependencies, too. It's not very clean, and rather complicated to port, but it's doable. 

Hmm... from my perspective, that sounds exactly like dependency management.  Well, they're there incase they're found useful.  I'm working on a project that's going to be in need of a way to build from source on any platform and these are the utilities that my programmer came back with.  Our use-case is sligthly different because we'll be needing the libraries on non-Linux as well, but having tried the `build.sh` this week first hand, I thought they'd help. :)

**Comment by probonopd** [comment]
Thanks @tresf I don't see at a moment _how_ this can help us but I am sure it will, sooner or later, be helpful in some way ;-)

**Comment by TheAssassin** [comment]
... most likely while trying to help another user packaging something as an AppImage.

**Comment by tresf** [comment]
>  I don't see at a moment how this can help us right now 

From what I'm reading in the source code of this project, many dependencies are manually fetched through methods such as tarballs, compiled and then utilized.  The utilities provided are intended to do just that, and through cmake, which seems to fulfill these two requests... 

* "It would be nice to have a modern and reliable build infrastructure."
* "a better handling of external dependencies."
* "build scripts are fragile" (#325)

It seems like these tools are upstream solutions to the mentioned problems (perhaps they don't provide the utilities yet but I don't see how that makes them unhelpful).

For example, with hunter... 

```cmake
hunter_add_package(fmt)
find_package(fmt CONFIG REQUIRED)
```

So an example of how this may be able to help is to spend more time upstream (e.g. PR an `inotify` recipe) and then less time downstream writing shell scripts.  This also helps other projects utilize the same tools, which in turn helps -- in part -- those tools gain wider adoption.  :)

**Comment by TheAssassin** [comment]
The problem is that we need to build these libraries in a special way before we can use them. None of these "package managers" can suffice that requirement any easier than just using what CMake gives us without such an external dependency. So, please don't try to over-engineer this, I don't really care whether to get the source from a tarball, or a source code repository, as long as we come up with a working and maintainable solution.

But if anyone cares enough to implement one of these options, we'd welcome a PR.

**Comment by TheAssassin** [comment]
We have had a CMake build infrastructure for a while already. Closing.
